File: src/pso/Particle.java
package pso;

import java.util.Random;
import java.util.function.Function;

public class Particle {

    private Vector position;

    private Vector velocity;

    private Vector bestKnownPosition;

    /**
     * Current particle's position
     * @return Returns copy of vector representing particles's position
     */
    public Vector getPosition() {
        return new Vector(this.position);
    }

    /**
     * Create particle with random position and velocity
     * @param dimension dimension in which particle exists, for instance dimension of fitness function
     * @param min lower bound of domain
     * @param max higher bound of domain
     * @return instance of particle
     */
    public static Particle createRandomParticle(int dimension, int min, int max) {
        var p = new Particle();
        p.initRandom(dimension, min, max);
        return p;
    }

    /**
     * Apply fintess function fn to current particle position in the given domain
     * @param fn Fitness function
     * @return Double which represents current fitness of particle
     */
    public Double apply(Function<Vector, Double> fn) {
        return fn.apply(position);
    }

    public void updateVelocity(double omega, double psi, Vector gBest) {
        Random random = new Random();
        this.velocity.map((i, vi) -> {
            var rp = random.nextDouble();
            var rg = random.nextDouble();
            return omega * vi + psi * rp * (this.bestKnownPosition.get(i) - this.position.get(i)) + omega * rg * (gBest.get(i)  - this.position.get(i));
        });
    }

    /**
     * Update particle's position according to it's actual velocity
     */
    public void updatePosition() {
        this.position.map((i, xi) -> {
            return xi + this.velocity.get(i);
        });
    }

    public Vector getBestPosition() {
       return bestKnownPosition;
    }

    public void updateBestPosition() {
        this.bestKnownPosition = new Vector(this.position);
    }

    private Particle() {}

    private void initRandom(int dimension, int min, int max) {
        this.position = Vector.random(dimension, min, max);
        this.velocity = Vector.random(dimension, (-1) * Math.abs(max - min), Math.abs(max - min));
        this.bestKnownPosition = new Vector(this.position);
    }
}

File: scala/src/main/scala/Particle.scala
import scala.collection.immutable.Vector
import scala.util.Random

object Particle {
  val random: Random.type = scala.util.Random
  def randomVector(dimension : Int, lower : Int, higher : Int): Vector[Double] = {
    (for (_ <- 0 until dimension) yield lower + (higher - lower) * random.nextDouble()).toVector
  }
}

class Particle(val fn : Vector[Double] => Double,
               val domain : Domain,
               val parameters: Parameters,
               var position: Vector[Double],
               var velocity : Vector[Double],
               var bestKnownPosition : Vector[Double],
               var bestKnownFitness : Double) {

  val random = new scala.util.Random()

  def this(domain : Domain,
           parameters : Parameters,
           fn : Vector[Double] => Double,
           position : Vector[Double],
           velocity : Vector[Double]) = {
    this(fn, domain, parameters, position, velocity, position, fn(position))
  }

  def this(dimension : Int, domain : Domain, parameters : Parameters, fn : Vector[Double] => Double) = {
    this(domain,
      parameters,
      fn,
      Particle.randomVector(dimension, domain.lowerBound, domain.higherBound),
      Particle.randomVector(dimension, domain.lowerBound, domain.higherBound))
  }

  def apply() : Double = {
    fn(position);
  }

  def updateVelocity(gBest : Vector[Double], omega : Double, phi_1 : Double, phi_2 : Double): Unit = {
    if(!domain.feasible(position)) {
      this.velocity = this.velocity.map(d => 0.002)
    }
    this.velocity = this.bestKnownPosition
      .lazyZip {
        this.position
      }
      .lazyZip(this.velocity)
      .lazyZip(gBest)
      .map((bestKnownPosition, position, velocity, gBest) => {
        omega * velocity + phi_1 * random.nextDouble() * (bestKnownPosition - position) + phi_2 * random.nextDouble() * (gBest - position)
      })
  }

  def updatePosition(): Unit = {
    this.position = this.position.lazyZip(this.velocity).map((position, velocity) => position + velocity)
  }

  def iterate(gBest : Vector[Double], iteration : Int) : Option[(Vector[Double], Double)] = {
    this.updateVelocity(gBest, parameters.getOmega(iteration), parameters.getPhi1, parameters.getPhi2)
    this.updatePosition()
    if(this.domain.feasible(this.position)) {
      val fitness = this.apply()
      if (fitness < 0) {
        println(this.position)
      }
      if(fitness < this.bestKnownFitness) {
        this.bestKnownPosition = this.position
        this.bestKnownFitness = fitness
        return Some(Tuple2[Vector[Double], Double](this.bestKnownPosition, this.bestKnownFitness))
      }
    }
    None
  }
}
